spirv_msl.hpp

```diff
@@ -317,6 +321,31 @@ class CompilerMSL : public CompilerGLSL
 		// May reduce performance in scenarios where arrays are copied around as value-types.
 		bool force_native_arrays = false;
 
+		// In a tessellation control shader, assume that more than one patch can be processed in a
+		// single workgroup. This requires changes to the way the InvocationId and PrimitiveId
+		// builtins are processed, but should result in more efficient usage of the GPU.
+		bool multi_patch_workgroup = false;
+
+		// If set, a vertex shader will be compiled as part of a tessellation pipeline.
+		// It will be translated as a compute kernel, so it can use the global invocation ID
+		// to index the output buffer.
+		bool vertex_for_tessellation = false;
+
+		typedef enum
+		{
+			NoIndex,
+			Index16,
+			Index32
+		} IndexType;
+
+		// The type of index in the index buffer, if present. For a compute shader, Metal
```

Comment:
I assume this only works because the PATCH primitive type does not allow primitive restart?

Reply:
Does Metal require this (does Metal actually support using a compute shader as a vertex shader?), or isn't it just our implementation that requires it?

Reply:
Looking at the code I'm only seeing the index buffer being used to resolve gl_VertexIndex, does Metal support using index buffers to guide compute kernels?

Reply:
> I assume this only works because the PATCH primitive type does not allow primitive restart?

Yep.



> Does Metal require this (does Metal actually support using a compute shader as a vertex shader?), or isn't it just our implementation that requires it?

It's just the current implementation that requires it, and only (as you later discovered) to resolve `gl_VertexIndex`. Note that you can't specify a compute function as the `vertexFunction` in a render pipeline, but you can do vertex-style stage input to a compute function. Ostensibly, this was intended to be used to run the vertex stage of a tessellation pipeline that was ported from D3D/OpenGL/Vulkan.

> Looking at the code I'm only seeing the index buffer being used to resolve gl_VertexIndex, does Metal support using index buffers to guide compute kernels?

Yes, but you have to specify the index buffer's location and index size when you create the pipeline object. (We had this issue with `tesc` shaders in the old implementation.)

Reply:
So to be clear, attribute fetching can be indexed by an index buffer in a Metal compute kernel? And that means the vertex shader will transparently unroll VBO + IBO -> to a flat control point array, which TESC then consumes?

Reply:
Yes.

Reply:
That's kinda cool actually, good to know.